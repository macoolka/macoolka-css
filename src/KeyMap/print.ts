/**
 * The provide getter from CssNode to string
 *
 * @getter
 */
import { indent } from 'mocoolka-fp/lib/format';
import {  PNode } from './Node';
// import { CssSelector} from './CssSelector';
import { toPairs, omit } from 'mocoolka-fp/lib/object';
import { Getter } from 'mocoolka-fp/lib/Monocle';
import {hyphenate} from 'mocoolka-fp/lib/string';
import {fromNullable} from 'mocoolka-fp/lib/Option';
const CRLF = '\n';
const SPACE = ' ';
const printValue = (a: string | number) => a;
const printName = (a: string) => hyphenate(a);

const printProperty = (i: number) => ([name, value]: [string, any]): string =>
    `${indent(i)}${printName(name)}:${SPACE}${printValue(value)};`;

const printSelectorNode = (i: number) =>
    <T>([name, prop]: [string, T]): string =>
        `${indent(i)}${printName(name)}${SPACE}{${CRLF}${printProperties(i + 1)(prop)}${CRLF}${indent(i)}}`;
const printMediaNode = (i: number) =>
        <T extends object>([name, prop]: [string,[string,T][]]): string =>
            `${indent(i)}${printName(name)}${SPACE}{${CRLF}${printSelector(i + 1)(prop)}${CRLF}${indent(i)}}`;
const printProperties = (i: number) => <T>(a: T): string => toPairs(a as {}).map(printProperty(i)).join(CRLF);
const printSelector = (i: number) =>
    <T extends object>(a: [string,T][]): string => a.map(printSelectorNode(i)).join(CRLF);
const printMedia = (i: number) =>
    <T extends object>(a: [string,[string,T][]][]): string => a.map(printMediaNode(i)).join(CRLF);
const printNode = (i: number) => <T extends object> (node: PNode<T>) => {
    const result = '';

  /*      const prop= omit(node,'selector');
       const selector=node.selector;
       result+=(selector===''||selector!==undefined)?
       printProperties(i)(prop):printProperties(i+1)({[`${selector}`]:prop}) */
      // result+=printProperties(i)(propC):`${CRLF}${printSelector(i)(selector)}`

    return `${printProperties(i)(omit(node, ['selector','media']))}${fromNullable(node.selector).
        map(a => `${CRLF}${printSelector(i)(a)}`).getOrElse('')}${fromNullable(node.media).
            map(a => `${CRLF}${printMedia(i)(a)}`).getOrElse('')}`;

    return result;
/*    return `${printProperties(i)(lensA.props.get(node))}${lensA.selector.getOption(node).
    map(a => `${CRLF}${printSelector(i)(a)}`).getOrElse('')}`; */
};

// export const printCss = (i: number = 0) => (css: CommonCss) => toPairs(css).map(printCssNode(i)).join(CRLF);
export const nodeToStringGetter = <T extends object>(i: number = 0) => new Getter<PNode<T>, string>(printNode(i));
export const toString = nodeToStringGetter(0).get;
