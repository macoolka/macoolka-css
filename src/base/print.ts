/**
 * The provide getter from CssNode to string
 *
 * @getter
 */
import { indent } from 'mocoolka-fp/lib/format';
import { CssNode, getProps, getSelector} from './CssNode';
import { CssSelector} from './CssSelector';
import { toPairs } from 'mocoolka-fp/lib/object';
import { Getter } from 'mocoolka-fp/lib/Monocle';
import {hyphenate} from 'mocoolka-fp/lib/string';
const CRLF = '\n';
const SPACE = ' ';
const printValue = (a: string | number) => a;
const printName = (a: string) => hyphenate(a);

const printProperty = (i: number) => ([name, value]: [string, any]): string =>
    `${indent(i)}${printName(name)}:${SPACE}${printValue(value)};`;

const printSelectorNode = (i: number) =>
    <T>([name, prop]: [string, T]): string =>
        `${indent(i)}${printName(name)}${SPACE}{${CRLF}${printProperties(i + 1)(prop)}${CRLF}${indent(i)}}`;

const printProperties = (i: number) => <T>(a: T): string => toPairs(a as {}).map(printProperty(i)).join(CRLF);
const printSelector = (i: number) =>
    <T>(a: CssSelector<T>): string => toPairs(a).map(printSelectorNode(i)).join(CRLF);
const printNode = (i: number) => <T> (node: CssNode<T>) =>
    `${printProperties(i)(getProps(node))}${getSelector(node).map(a => `${CRLF}${printSelector(i)(a)}`).getOrElse('')}`;
// export const printCss = (i: number = 0) => (css: CommonCss) => toPairs(css).map(printCssNode(i)).join(CRLF);
export const cssNodeToStringGetter = <T>(i: number = 0) => new Getter<CssNode<T>, string>(printNode(i));
export const toString = cssNodeToStringGetter(0).get;
