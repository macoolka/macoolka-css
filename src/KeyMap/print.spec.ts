import { nodeToStringGetter } from './print';
const p1 = {
    color: 'red',
    marginTop: '12px',
    marginLeft: '1px',
};
const selector =
{
    '$:hover': {
        color: 'green',
        marginTop: '20px',
        marginLeft: '21px',
    },
    '$:focus': {
        color: 'blue',
        marginTop: '2px',
        marginLeft: '3px',
    }
}

describe('css print', () => {
    it('commonToStringGetter', () => {
        expect(nodeToStringGetter<any>().get({...p1,selector})).toMatchSnapshot();
    });
});