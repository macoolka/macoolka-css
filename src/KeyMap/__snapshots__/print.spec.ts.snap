// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css print commonToStringGetter 1`] = `
"color: red;
margin-top: 12px;
margin-left: 1px;
$:hover {
  color: green;
  margin-top: 20px;
  margin-left: 21px;
}
$:focus {
  color: blue;
  margin-top: 2px;
  margin-left: 3px;
}"
`;
